Mathlib/Algebra/Quaternion.lean

```diff
@@ -479,126 +495,133 @@ instance instRing : Ring ℍ[R,c₁,c₂] where
   mul_one _ := by ext <;> simp
 
 @[norm_cast, simp]
-theorem coe_mul : ((x * y : R) : ℍ[R,c₁,c₂]) = x * y := by ext <;> simp
-
--- TODO: add weaker `MulAction`, `DistribMulAction`, and `Module` instances (and repeat them
--- for `ℍ[R]`)
-instance [CommSemiring S] [Algebra S R] : Algebra S ℍ[R,c₁,c₂] where
-  smul := (· • ·)
-  algebraMap :=
-  { toFun s := coe (algebraMap S R s)
-    map_one' := by simp only [map_one, coe_one]
-    map_zero' := by simp only [map_zero, coe_zero]
-    map_mul' x y := by simp only [map_mul, coe_mul]
-    map_add' x y := by simp only [map_add, coe_add] }
-  smul_def' s x := by ext <;> simp [Algebra.smul_def]
+theorem coe_mul : ((x * y : R) : ℍ[R,c₁,c₂,c₃]) = x * y := by ext <;> simp
+
+instance [Monoid S] [MulAction S R] : MulAction S ℍ[R,c₁,c₂,c₃] where
+  one_smul _ := by ext <;> simp
+  mul_smul _ _ _ := by ext <;> simp [mul_smul]
+
+instance [Monoid S] [DistribMulAction S R] : DistribMulAction S ℍ[R,c₁,c₂,c₃] where
+  smul_add _ _ _ := by ext <;> simp [smul_add]
+  smul_zero _ := by ext <;> simp [smul_zero]
+
+instance [Semiring S] [Module S R] : Module S ℍ[R,c₁,c₂,c₃] where
+  add_smul _ _ _ := by ext <;> simp [add_smul]
+  zero_smul _ := by ext <;> simp [zero_smul]
+
+instance [CommSemiring S] [Algebra S R] : Algebra S ℍ[R,c₁,c₂,c₃] where
+  algebraMap := {
+    toFun s := coe (algebraMap S R s)
+    map_one' := by simp
+    map_mul' := by simp
+    map_zero' := by simp
+    map_add' := by simp
+  }
   commutes' s x := by ext <;> simp [Algebra.commutes]
+  smul_def' s x := by ext <;> simp [Algebra.smul_def]
```

Comment:
this PR seems to be mixing two things; can you make this generalization in a separate PR?